src/main/java/com/zendesk/maxwell/MaxwellParser.java

```diff
@@ -133,71 +126,122 @@ private MaxwellAbstractRowsEvent processRowsEvent(AbstractRowEvent e) throws Sch
-
-	}
-
-	public MaxwellAbstractRowsEvent getEvent(boolean stopAtNextTableMap) throws Exception {
-        BinlogEventV4 v4Event;
-        MaxwellAbstractRowsEvent event;
-		while (true) {
-			v4Event = queue.poll(100, TimeUnit.MILLISECONDS);
-
-			if ( v4Event == null )
-				return null;
-
-			switch(v4Event.getHeader().getEventType()) {
 			case MySQLConstants.WRITE_ROWS_EVENT:
+				ew = new MaxwellWriteRowsEvent((WriteRowsEvent) e, table, filter);
+				break;
 			case MySQLConstants.WRITE_ROWS_EVENT_V2:
+				ew = new MaxwellWriteRowsEvent((WriteRowsEventV2) e, table, filter);
+				break;
 			case MySQLConstants.UPDATE_ROWS_EVENT:
+				ew = new MaxwellUpdateRowsEvent((UpdateRowsEvent) e, table, filter);
+				break;
 			case MySQLConstants.UPDATE_ROWS_EVENT_V2:
+				ew = new MaxwellUpdateRowsEvent((UpdateRowsEventV2) e, table, filter);
+				break;
 			case MySQLConstants.DELETE_ROWS_EVENT:
+				ew = new MaxwellDeleteRowsEvent((DeleteRowsEvent) e, table, filter);
+				break;
 			case MySQLConstants.DELETE_ROWS_EVENT_V2:
-				rowEventsProcessed++;
-				event = processRowsEvent((AbstractRowEvent) v4Event);
-				if ( event.matchesFilter() )
-					return event;
+				ew = new MaxwellDeleteRowsEvent((DeleteRowsEventV2) e, table, filter);
 				break;
-			case MySQLConstants.TABLE_MAP_EVENT:
-				if ( stopAtNextTableMap)
-					return null;
+			default:
+				return null;
+		}
+		return ew;
+	}
 
-				tableCache.processEvent(this.schema, (TableMapEvent) v4Event);
-				break;
-			case MySQLConstants.QUERY_EVENT:
-				processQueryEvent((QueryEvent) v4Event);
+	private LinkedList<MaxwellAbstractRowsEvent> getTransactionEvents() throws Exception {
+		BinlogEventV4 v4Event;
+		MaxwellAbstractRowsEvent event;
+
+		LinkedList<MaxwellAbstractRowsEvent> list = new LinkedList<>();
+
+		// currently to satisfy the test interface, the contract is to return null
+		// if the queue is empty.  should probably just replace this with an optional timeout...
+
+		while ( true ) {
+			v4Event = queue.poll(100, TimeUnit.MILLISECONDS);
+			if (v4Event == null)
+				continue;
+
+			switch(v4Event.getHeader().getEventType()) {
+				case MySQLConstants.WRITE_ROWS_EVENT:
+				case MySQLConstants.WRITE_ROWS_EVENT_V2:
+				case MySQLConstants.UPDATE_ROWS_EVENT:
+				case MySQLConstants.UPDATE_ROWS_EVENT_V2:
+				case MySQLConstants.DELETE_ROWS_EVENT:
+				case MySQLConstants.DELETE_ROWS_EVENT_V2:
+					rowEventsProcessed++;
+					event = processRowsEvent((AbstractRowEvent) v4Event);
+
+					if ( event.matchesFilter() )
+						list.add(event);
+					break;
+				case MySQLConstants.TABLE_MAP_EVENT:
+					tableCache.processEvent(this.schema, (TableMapEvent) v4Event);
+					break;
+				case MySQLConstants.QUERY_EVENT:
+					QueryEvent qe = (QueryEvent) v4Event;
+					// TODO: need to handle this case.
+					// The MySQL guys say some storage engines will output a "COMMIT" QUERY_EVENT at the
+					// end of the stream.
+					throw new RuntimeException("Unhandled QueryEvent: " + qe);
```

Comment:
so trying out this branch in dev has found a few of us getting these messages from our dockerized mysql (and a few weirdly not) and maxwell subsequently crashing. I'm a bit fuzzy on my mysql internals knowledge, what should we do here?